File: samples/csharp_dotnetcore/23.facebook-events/Bots/FacebookBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Schema;
using Microsoft.BotBuilderSamples.FacebookModel;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples.Bots
{
    public class FacebookBot : ActivityHandler
    {
        // These are the options provided to the user when they message the bot
        const string FacebookPageIdOption = "Facebook Id";
        const string QuickRepliesOption = "Quick Replies";
        const string PostBackOption = "PostBack";

        protected readonly ILogger Logger;

        public FacebookBot(ILogger<FacebookBot> logger)
        {
            Logger = logger;
        }

        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            Logger.LogInformation("Processing a Message Activity.");

            // Show choices if the Facebook Payload from ChannelData is not handled
            if (!await ProcessFacebookPayload(turnContext, turnContext.Activity.ChannelData, cancellationToken))
                await ShowChoices(turnContext, cancellationToken);
        }

        protected override async Task OnEventActivityAsync(ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
        {
            Logger.LogInformation("Processing an Event Activity.");

            // Analyze Facebook payload from EventActivity.Value
            await ProcessFacebookPayload(turnContext, turnContext.Activity.Value, cancellationToken);
        }

        private static async Task ShowChoices(ITurnContext turnContext, CancellationToken cancellationToken)
        {
            // Create choices for the prompt
            var choices = new List<Choice>();
            choices.Add(new Choice() { Value = QuickRepliesOption, Action = new CardAction() { Title = QuickRepliesOption, Type = ActionTypes.PostBack, Value = QuickRepliesOption } });
            choices.Add(new Choice() { Value = FacebookPageIdOption, Action = new CardAction() { Title = FacebookPageIdOption, Type = ActionTypes.PostBack, Value = FacebookPageIdOption } });
            choices.Add(new Choice() { Value = PostBackOption, Action = new CardAction() { Title = PostBackOption, Type = ActionTypes.PostBack, Value = PostBackOption } });

            // Create the prompt message
            var message = ChoiceFactory.ForChannel(turnContext.Activity.ChannelId, choices, "What Facebook feature would you like to try? Here are some quick replies to choose from!");
            await turnContext.SendActivityAsync(message, cancellationToken);
        }

        private async Task<bool> ProcessFacebookPayload(ITurnContext turnContext, object data, CancellationToken cancellationToken)
        {
            // At this point we know we are on Facebook channel, and can consume the Facebook custom payload
            // present in channelData.
            try
            {
                var facebookPayload = (data as JObject)?.ToObject<FacebookPayload>();

                if (facebookPayload != null)
                {
                    // PostBack
                    if (facebookPayload.PostBack != null)
                    {
                        await OnFacebookPostBack(turnContext, facebookPayload.PostBack, cancellationToken);
                        return true;
                    }

                    // Optin
                    else if (facebookPayload.Optin != null)
                    {
                        await OnFacebookOptin(turnContext, facebookPayload.Optin, cancellationToken);
                        return true;
                    }

                    // Quick reply
                    else if (facebookPayload.Message?.QuickReply != null)
                    {
                        await OnFacebookQuickReply(turnContext, facebookPayload.Message.QuickReply, cancellationToken);
                        return true;
                    }

                    // Echo
                    else if (facebookPayload.Message?.IsEcho != null && facebookPayload.Message.IsEcho)
                    {
                        await OnFacebookEcho(turnContext, facebookPayload.Message, cancellationToken);
                        return true;
                    }

                    // TODO: Handle other events that you're interested in...
                }
            }
            catch(JsonSerializationException e)
            {
                if (turnContext.Activity.ChannelId != Bot.Connector.Channels.Facebook)
                {
                    await turnContext.SendActivityAsync("This sample is intended to be used with a Facebook bot.");
                }
                else
                {
                    throw;
                }
            }
          
            return false;
        }

        protected virtual async Task OnFacebookOptin(ITurnContext turnContext, FacebookOptin optin, CancellationToken cancellationToken)
        {
            Logger.LogInformation("Optin message received.");

            // TODO: Your optin event handling logic here...
        }

        protected virtual async Task OnFacebookEcho(ITurnContext turnContext, FacebookMessage facebookMessage, CancellationToken cancellationToken)
        {
            Logger.LogInformation("Echo message received.");

            // TODO: Your echo event handling logic here...
        }

        protected virtual async Task OnFacebookPostBack(ITurnContext turnContext, FacebookPostback postBack, CancellationToken cancellationToken)
        {
            Logger.LogInformation("PostBack message received.");

            // TODO: Your PostBack handling logic here...

            // Answer the postback, and show choices
            var reply = MessageFactory.Text("Are you sure?");
            await turnContext.SendActivityAsync(reply, cancellationToken);
            await ShowChoices(turnContext, cancellationToken);
        }

        protected virtual async Task OnFacebookQuickReply(ITurnContext turnContext, FacebookQuickReply quickReply, CancellationToken cancellationToken)
        {
            Logger.LogInformation("QuickReply message received.");

            // TODO: Your quick reply event handling logic here...

            // Process the message by checking the Activity.Text.  The FacebookQuickReply could also contain a json payload.

            // Initially the bot offers to showcase 3 Facebook features: Quick replies, PostBack and getting the Facebook Page Name.
            switch (turnContext.Activity.Text)
            {
                // Here we showcase how to obtain the Facebook page id.
                // This can be useful for the Facebook multi-page support provided by the Bot Framework.
                // The Facebook page id from which the message comes from is in turnContext.Activity.Recipient.Id.
                case FacebookPageIdOption:
                    {
                        var reply = MessageFactory.Text($"This message comes from the following Facebook Page: {turnContext.Activity.Recipient.Id}");
                        await turnContext.SendActivityAsync(reply, cancellationToken);
                        await ShowChoices(turnContext, cancellationToken);

                        break;
                    }

                // Here we send a HeroCard with 2 options that will trigger a Facebook PostBack.
                case PostBackOption:
                    {
                        var card = new HeroCard
                        {
                            Text = "Is 42 the answer to the ultimate question of Life, the Universe, and Everything?",
                            Buttons = new List<CardAction>
                                    {
                                        new CardAction() { Title = "Yes", Type = ActionTypes.PostBack, Value = "Yes" },
                                        new CardAction() { Title = "No", Type = ActionTypes.PostBack, Value = "No" },
                                    },
                        };
                        
                        var reply = MessageFactory.Attachment(card.ToAttachment());
                        await turnContext.SendActivityAsync(reply, cancellationToken);

                        break;
                    }

                // By default we offer the users different actions that the bot supports, through quick replies.
                case QuickRepliesOption:
                default:
                    {
                        await ShowChoices(turnContext, cancellationToken);

                        break;
                    }
            }
        }
    }
}

File: samples/python/23.facebook-events/bots/facebook_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs.choices import Choice, ChoiceFactory
from botbuilder.core import ActivityHandler, MessageFactory, TurnContext, CardFactory
from botbuilder.schema import ChannelAccount, CardAction, ActionTypes, HeroCard

FACEBOOK_PAGEID_OPTION = "Facebook Id"
QUICK_REPLIES_OPTION = "Quick Replies"
POSTBACK_OPTION = "PostBack"


class FacebookBot(ActivityHandler):
    async def on_members_added_activity(
        self, members_added: [ChannelAccount], turn_context: TurnContext
    ):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity("Hello and welcome!")

    async def on_message_activity(self, turn_context: TurnContext):
        if not await self._process_facebook_payload(
            turn_context, turn_context.activity.channel_data
        ):
            await self._show_choices(turn_context)

    async def on_event_activity(self, turn_context: TurnContext):
        await self._process_facebook_payload(turn_context, turn_context.activity.value)

    async def _show_choices(self, turn_context: TurnContext):
        choices = [
            Choice(
                value=QUICK_REPLIES_OPTION,
                action=CardAction(
                    title=QUICK_REPLIES_OPTION,
                    type=ActionTypes.post_back,
                    value=QUICK_REPLIES_OPTION,
                ),
            ),
            Choice(
                value=FACEBOOK_PAGEID_OPTION,
                action=CardAction(
                    title=FACEBOOK_PAGEID_OPTION,
                    type=ActionTypes.post_back,
                    value=FACEBOOK_PAGEID_OPTION,
                ),
            ),
            Choice(
                value=POSTBACK_OPTION,
                action=CardAction(
                    title=POSTBACK_OPTION,
                    type=ActionTypes.post_back,
                    value=POSTBACK_OPTION,
                ),
            ),
        ]

        message = ChoiceFactory.for_channel(
            turn_context.activity.channel_id,
            choices,
            "What Facebook feature would you like to try? Here are some quick replies to choose from!",
        )
        await turn_context.send_activity(message)

    async def _process_facebook_payload(self, turn_context: TurnContext, data) -> bool:
        if "postback" in data:
            await self._on_facebook_postback(turn_context, data["postback"])
            return True

        if "optin" in data:
            await self._on_facebook_optin(turn_context, data["optin"])
            return True

        if "message" in data and "quick_reply" in data["message"]:
            await self._on_facebook_quick_reply(
                turn_context, data["message"]["quick_reply"]
            )
            return True

        if "message" in data and data["message"]["is_echo"]:
            await self._on_facebook_echo(turn_context, data["message"])
            return True

    async def _on_facebook_postback(
        self, turn_context: TurnContext, facebook_postback: dict
    ):
        # TODO: Your PostBack handling logic here...

        reply = MessageFactory.text(f"Postback: {facebook_postback}")
        await turn_context.send_activity(reply)
        await self._show_choices(turn_context)

    async def _on_facebook_quick_reply(
        self, turn_context: TurnContext, facebook_quick_reply: dict
    ):
        # TODO: Your quick reply event handling logic here...

        if turn_context.activity.text == FACEBOOK_PAGEID_OPTION:
            reply = MessageFactory.text(
                f"This message comes from the following Facebook Page: {turn_context.activity.recipient.id}"
            )
            await turn_context.send_activity(reply)
            await self._show_choices(turn_context)
        elif turn_context.activity.text == POSTBACK_OPTION:
            card = HeroCard(
                text="Is 42 the answer to the ultimate question of Life, the Universe, and Everything?",
                buttons=[
                    CardAction(title="Yes", type=ActionTypes.post_back, value="Yes"),
                    CardAction(title="No", type=ActionTypes.post_back, value="No"),
                ],
            )
            reply = MessageFactory.attachment(CardFactory.hero_card(card))
            await turn_context.send_activity(reply)
        else:
            print(facebook_quick_reply)
            await turn_context.send_activity("Quick Reply")
            await self._show_choices(turn_context)

    async def _on_facebook_optin(self, turn_context: TurnContext, facebook_optin: dict):
        # TODO: Your optin event handling logic here...
        print(facebook_optin)
        await turn_context.send_activity("Opt In")

    async def _on_facebook_echo(
        self, turn_context: TurnContext, facebook_message: dict
    ):
        # TODO: Your echo event handling logic here...
        print(facebook_message)
        await turn_context.send_activity("Echo")
